src/System.Device.Gpio.Tests/GpioControllerTestBase.cs

```diff
@@ -153,7 +150,81 @@ void callback(object sender, PinValueChangedEventArgs pinValueChangedEventArgs)
+            {
+                controller.OpenPin(InputPin, PinMode.Input);
+                controller.OpenPin(OutputPin, PinMode.Output);
+                controller.Write(OutputPin, PinValue.Low);
+
+                controller.RegisterCallbackForPinValueChangedEvent(InputPin, PinEventTypes.Rising, (o, e) => {
+                    risingEventOccuredCount++;
+                });
+                                
+                controller.RegisterCallbackForPinValueChangedEvent(InputPin, PinEventTypes.Rising, callback);
+                controller.RegisterCallbackForPinValueChangedEvent(InputPin, PinEventTypes.Rising, (o, e) => {
+                    risingEventOccuredCount++;
+                    if (fallingEventOccuredCount == 4)
+                    {
+                        controller.UnregisterCallbackForPinValueChangedEvent(InputPin, callback);
+                    }
+                });
+
+                controller.RegisterCallbackForPinValueChangedEvent(InputPin, PinEventTypes.Falling, (o, e) => {
+                    fallingEventOccuredCount++;
+                });
+
+                for (int i = 0; i < 10; i++)
+                {
+                    controller.Write(OutputPin, PinValue.High);
+                    Thread.Sleep(10);
+                    controller.Write(OutputPin, PinValue.Low);
+                    Thread.Sleep(10);
+                }
+                Assert.Equal(25, risingEventOccuredCount);
+                Assert.Equal(10, fallingEventOccuredCount);
+
+                void callback(object sender, PinValueChangedEventArgs e)
+                {
+                    risingEventOccuredCount++;
+                }
+            }
+        }
+
+        [Fact]
+        public void WaitForEventTest()
+        {
+            using (GpioController controller = new GpioController(GetTestNumberingScheme(), GetTestDriver()))
+            {
+                bool cancelled = false;
+                CancellationTokenSource tokenSource = new CancellationTokenSource();
+                controller.OpenPin(InputPin, PinMode.Input);
+                controller.OpenPin(LedPin, PinMode.Input);
+                controller.OpenPin(OutputPin, PinMode.Output);
+                controller.Write(OutputPin, PinValue.Low);
+
+                controller.RegisterCallbackForPinValueChangedEvent(InputPin, PinEventTypes.Rising, (o, e) =>
+                {
+                        tokenSource.Cancel();
+                        cancelled = true;
+                });
+
+                Task.Run(() =>
+                {
+                    controller.Write(OutputPin, PinValue.High);
+                    Thread.Sleep(10);
+                    controller.Write(OutputPin, PinValue.Low);
+                });
+
+                WaitForEventResult result = controller.WaitForEvent(LedPin, PinEventTypes.Rising, tokenSource.Token);
+                Assert.True(cancelled);
+                Assert.True(result.TimedOut);
+            }
+        }
```

Comment:
other scenarios which might be worth testing:
- register event falling/rising and trigger opposite event and expect the callback not to be called
- register few events, unregister them all and expect not to be called
- register, unregister, register again - expect only 1 event
- WaitForEvent falling
- cancellation token scenarios

Reply:
Yeah, there could be many scenarios, tried to test at least these to work, will add more

Reply:
added more scenarios